solutions/usaco-917.mdx

```diff
@@ -0,0 +1,124 @@
+
+Predicting the range of possible initial flows is similar and essentially symmetric, where we scan backwards and keep track of a working range $[a,b]$ that is narrowed / modified appropriately by each highway feature.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <fstream>
+using namespace std;
+     
+int main(void)
+{
+  int N, A[100], B[100];
+  string T[100];
+     
+  ifstream fin ("traffic.in");
+  fin >> N;
+  for (int i=0; i<N; i++) fin >> T[i] >> A[i] >> B[i];
+     
+  ofstream fout ("traffic.out");  
+     
+  int a = -999999999, b = 999999999;
+  for (int i=N-1; i>=0; i--) {
+    if (T[i] == "none") { a = max(a, A[i]); b = min(b, B[i]); }
+    if (T[i] == "off") { a += A[i]; b += B[i]; }
+    if (T[i] == "on") { a -= B[i]; b -= A[i]; a = max(0,a); }
+  }
+  fout << a << " " << b << "\n";
+     
+  a = -999999999, b = 999999999;
+  for (int i=0; i<N; i++) {
+    if (T[i] == "none") { a = max(a, A[i]); b = min(b, B[i]); }
+    if (T[i] == "on") { a += A[i]; b += B[i]; }
+    if (T[i] == "off") { a -= B[i]; b -= A[i]; a = max(0,a); }
+  }
+  fout << a << " " << b << "\n";
+ 
+  return 0;
+}
+```
+
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.StringTokenizer;
+
+public class traffic {
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new FileReader("traffic.in"));
+        int N = Integer.parseInt(br.readLine());
+        int [] A = new int [100];
+        int [] B = new int [100];
+        String [] T = new String [100];
+        for (int i = 0; i < N; i++) {
+            StringTokenizer s = new StringTokenizer(br.readLine());
+            T[i] = s.nextToken();
+            A[i] = Integer.parseInt(s.nextToken());
+            B[i] = Integer.parseInt(s.nextToken());
+        }
+        PrintWriter out = new PrintWriter("traffic.out");
+
+        // Below is the initial range (a,b).  We will narrow/modify this range as we scan through
+        // the highway components from mile N to mile 1.  The resulting range will be the range of
+        // possible initial traffic flows.
+        int a = -999999999, b = 999999999;
+
+        // Note that we are scanning backwards from mile N to mile 1 here.
+        for (int i=N-1; i>=0; i--) {
+            // if there is no ramp, the possible range (a,b) is narrowed to the range given by the sensor.
+            if (T[i].equals("none")) { a = Math.max(a, A[i]); b = Math.min(b, B[i]); }
```

Comment:
put each statement on their own line just like in the c++ version

Reply:
see C++ version -- I don't think this is necessary, though I'm fine eithe rway. I do think that lines 94/97 should have the whitespace after `T[i]` removed